### PR TITLE
Add support for system ID of position controller

### PR DESCRIPTION
**Pull request by bnsgeyer**
This PR adds additional inputs to conduct System ID on the position controller.  It adds the following inputs
longitudinal velocity (body axis fwd/aft)
lateral velocity (body axis left/right)
position disturbance in body longitudinal axis
position disturbance in body lateral axis
velocity disturbance in body longitudinal axis
velocity disturbance in body lateral axis

this also includes the PSC logging during the system ID at the system ID logging rate.  It allows the pilot to make velocity command inputs over top of the axis being swept but it is debatable whether this is even necessary since it should hold position except for the axis being swept.

so far i have tested this in SITL. would like @lthall to look at it.  Not getting the expected response for the pilot velocity commanded inputs.

### CODE REVIEW

**Comment by bnsgeyer** [comment]
@lthall could you review this to make sure I did this right. I have flown this in a heli and got velocity drb results.

**Comment by lthall** [comment]
This all looks great but using loiter like this is over kill I think.

I would suggest using the simple reposition code or just holding position without human input. People can always switch out of sysid mode.

https://github.com/ArduPilot/ardupilot/blob/c1831bae320c00338307c7c5f15e1b6b6b50abcc/ArduCopter/mode.cpp#L759C33-L759C33

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp:236]: @lthall why are you adding a vel correction here?  I thought we decided that the pilot would not have any control over the position or velocity while in the mode.

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp:375]: This needs to include input_vel from the chirp signal

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp]: So we don't have to update the pilot's desired accelerations?  I thought I would need to set this to zero

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [inline ArduCopter/mode_systemid.cpp:416]: Do we end up with double logs if PID logging is on already?

File: ArduCopter/mode_systemid.cpp
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp:15]: your choice of course but I'd remove the "Measured" from the front of some of these... I'm just not sure the extra specificity is required.

Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp:15]: I am being very specific because the sweep is being added to the measured velocity or position and not to the commanded(i.e. Input).  So I think this is necessary.  @lthall do you have any preference on this?

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp:416]: No.  this is independently logged for System ID mode

File: ArduCopter/mode.h
Comment (bnsgeyer) [reply ArduCopter/mode.h:1633]: answered at devcall.  closing comment

File: ArduCopter/Attitude.cpp
Comment (bnsgeyer) [reply ArduCopter/Attitude.cpp]: done. thanks!

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp]: removed.  thanks!

**Comment by bnsgeyer** [comment]
@rmackay9  functionally checked this in SITL and have conducted several sysid flights on a real heli using the position controller sweep settings.  Everything checks out.  Ready to be merged.

**Comment by bnsgeyer** [comment]
@lthall  and @IamPete1  could you guys give an approval on this.

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [inline ArduCopter/mode_systemid.cpp:100]: Its not really related to this PR, but it looks to me like we can get left on the acro stab col when not in acro.....

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [inline ArduCopter/mode_systemid.cpp:84]: I don't think the extra loiter check here is doing anything? `has_manual_throttle` returns false for loiter. So `!copter.flightmode->has_manual_throttle()` will already be true so there is no need to `|| true`?

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [inline ArduCopter/mode_systemid.cpp:165]: Could this whole section be put in to a new function? So we would have `ModeSystemId::run_pos_control()` and `ModeSystemId::run_angle_rate_mixer_control()` (I'm sure there is a better name we could come up with.) Then this core `run` function would just call into one or the other based on `axis`.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:648]: I think `comb_pos.xy() += _disturb_pos;` will work.

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [reply ArduCopter/mode_systemid.cpp:100]: Maybe it only matters that its right in acro, stabilize and sys ID?

Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp:100]: @IamPete1 This only applies to acro_heli and stabilize_heli flight modes.  It is specified for the init of each of these modes.  Beings that SysID for axes 1-13 uses stabilize mode, this is set to true for heli.

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp:84]: Good catch.  This must be a left over from early attempts to keep it from switching into system ID for axes 1-13 while in loiter.  I have captured that behavior in the sections below.

This really needs to capture that a user can't switch into system ID if they aren't flying.  For heli's that can be quite dangerous and I don't think multicopters want to takeoff with the aircraft doing sysid.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp
Comment (bnsgeyer) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:648]: I only pass in an xy disturbance, not xyz.  So I have to breakout the assignment to the 3 axis vector.

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp:165]: Some of this happens prior to the SID_AXIS case statement because the targets are determined for the stabilize flight mode from the pilot's inputs.  Not sure I want to mess with this.  I could make the if statement conditional look cleaner

Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp]: I cleaned up the conditional statement to make it look cleaner and read easier.  I'm don't think it is wise to break this out as @IamPete1 suggests.

**Comment by bnsgeyer** [comment]
I've tested this in a real heli.  No issues.  @IamPete1, would you approve please.  I plan to put this on tomorrow's devcall unless @rmackay9 is happy to merge once it is approved.
Thanks!

File: ArduCopter/mode_systemid.cpp
Comment (IamPete1) [inline ArduCopter/mode_systemid.cpp:433]: This would be neater as a switch statement.

File: ArduCopter/mode.h
Comment (IamPete1) [inline ArduCopter/mode.h:6]: I think this could just be included in the cpp.

**Review by IamPete1 — APPROVED** [review]
A few minor things but basically looks good.

File: ArduCopter/mode.h
Comment (bnsgeyer) [reply ArduCopter/mode.h]: Done. Thanks

File: ArduCopter/mode_systemid.cpp
Comment (bnsgeyer) [reply ArduCopter/mode_systemid.cpp]: Done. Thanks

File: libraries/AC_AttitudeControl/AC_PosControl.cpp
Comment (bnsgeyer) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: Done. Thanks

File: libraries/AC_AttitudeControl/AC_PosControl.h
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.h:459]: it would be nice to have comments besides these variables like the ones above it.

File: ArduCopter/mode.h
Comment (lthall) [inline ArduCopter/mode.h:1634]: Loiter isn't the best name

**Review by lthall — APPROVED** [review]

File: libraries/AC_AttitudeControl/AC_PosControl.h
Comment (bnsgeyer) [reply libraries/AC_AttitudeControl/AC_PosControl.h]: done.  Thanks!

File: ArduCopter/mode.h
Comment (bnsgeyer) [reply ArduCopter/mode.h]: Done. Thanks!

**Comment by bnsgeyer** [comment]
Here is the technical paper that this PR supported.
[OuterLoop_HQs_Paper_VFS_2024_final.pdf](https://github.com/ArduPilot/ardupilot/files/15300915/OuterLoop_HQs_Paper_VFS_2024_final.pdf)

**Pull request merged**
